### PR TITLE
IPP Feedback Banner - final PR

### LINKED ISSUE #8116: :sparkles: IPP in-app feedback banner
IPP in-app feedback banner redirecting to an online survey.

The goal of the project is to encourage specific groups of merchants to give feedback on their needs related to IPP. We want to identify 3 specific user groups and display a banner requesting feedback:
1. Merchants who have just enabled the COD feature and have not yet completed any IPP transaction.
2. Merchants who have completed at least 1, but less than 10 transactions in the last 30 days.
3. Merchants who are actively collecting IPP, and have completed more than 10 transactions in the last 30 days.

More details in the project thread: pdfdoF-20T-p2

- [x] Integrating `/payments/transactions/summary` endpoint to FluxC - https://github.com/wordpress-mobile/WordPress-FluxC-Android/issues/2625
- [x] Business logic identifying a specific user group
- [x] Storing "banner dismissed" event
- [x] Storing "survey completed" event
- [x] Business logic deciding if the banner should be displayed to the current user
- [x] Feature flag
- [x] Banner UI
- [x] Prioritise IPP banner over "Create orders" banner
- [x] Update banner display logic to meet new criteria
- [x] Ensure all the analytics events are tracked

### PR DESCRIPTION
**Pull request by samiuelson**
### IPP Feedback Banner - Project's final PR

This is a complete feature PR consisting of the following partial PRs, reviewed and merged to the this feature branch:
1. https://github.com/woocommerce/woocommerce-android/pull/8159
2. https://github.com/woocommerce/woocommerce-android/pull/8165
3. https://github.com/woocommerce/woocommerce-android/pull/8180
4. https://github.com/woocommerce/woocommerce-android/pull/8205
5. https://github.com/woocommerce/woocommerce-android/pull/8225
6. https://github.com/woocommerce/woocommerce-android/pull/8230

Closes: https://github.com/woocommerce/woocommerce-android/issues/8116

### Description
The goal of the project is to collect feedback from IPP users. We target three groups of merchants and show dedicated banners and feedback surveys.

- Specs for the banner title and message texts: p1674116096400209/1674112248.357019-slack-C0354HSNUJH

- More details in the project thread: pdfdoF-20T-p2

### Testing instructions

The banner should be visible on the orders page. The condition required to see the banner is to have COD enabled. 
If the user has ever completed IPP transactions banner type 2. or 3. (more than 10 transactions in the last 30 days) is shown. If they haven't made any IPP transactions banner 1. is shown.

### Images
The banner is placed on the Orders screen. Three banner types are shown in the table below.

| 1. IPP newbie users  | 2. IPP beginners | 3. IPP active users |
| ------------- | ------------- | ------------- | 
| Screenshot_20230123-113541  | Screenshot_20230123-113713  | Screenshot_20230123-113834 |

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListViewModel.kt [hunk 1]
```diff
@@ -161,6 +173,16 @@ class OrderListViewModel @Inject constructor(
                 )
             }
         }
+
+        viewModelScope.launch {
+            val shouldShowFeedbackBanner = FeatureFlag.IPP_FEEDBACK_BANNER.isEnabled() && shouldShowFeedbackBanner()
+            if (shouldShowFeedbackBanner) {
+                val bannerData = getIPPFeedbackBannerData()
+                viewState = viewState.copy(ippFeedbackBannerState = IPPSurveyFeedbackBannerState.Visible(bannerData))
+                trackIPPBannerEvent(AnalyticsEvent.IPP_FEEDBACK_BANNER_SHOWN)
```
Comment (AnirudhBhat) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListViewModel.kt:187]: ⚠️ I'm not seeing unit tests for any of the tracking related to the IPP feedback survey. We might want to add it. WDYT?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/feedback/ipp/GetIPPFeedbackBannerData.kt [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        }
+    }
+
+    private fun requireTransactionsCount(response: WooPayload<WCPaymentTransactionsSummaryResult>): Int {
+        return checkNotNull(response.result?.transactionsCount) { "Transactions count must not be null" }
+    }
+
+    private fun requireSuccessfulTransactionsSummaryResponse(response: WooPayload<WCPaymentTransactionsSummaryResult>) {
+        if (response.isError || response.result == null) {
+            throw IllegalStateException("Failed to fetch transactions summary")
```
Comment (AnirudhBhat) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/feedback/ipp/GetIPPFeedbackBannerData.kt]: ❓ I'm not sure if throwing an exception when the API call fails is the right way here. I just experienced this on my testing site. The transactions summary API was returning a 500 error and my app was crashing.

I'm thinking we should just log the error instead of throwing it in other places in this file as well. Can we log the error with the cause instead of crashing? I think won't letting the user use the app because of an API failure on the feedback banner is not a good UX since ipp feedback survey is not a must for users to use the app..

 WDYT?

**Review by AnirudhBhat — COMMENTED** [review]
Thanks for your work on this @samiuelson. Great work!

I've smoke tested the app and it seems to be working fine. I couldn't do thorough testing today because of a migraine. I'll be AFK tomorrow and will be back on Friday. But it will be late for your PR merge.

Since I've already reviewed major part of the PR, I've skimmed through the code and it looks good.

I'll leave it to @vanessamaartinstt for testing the app.

I've left a couple of questions to think about. Let me know what you feel about those.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/feedback/ipp/GetIPPFeedbackBannerData.kt (hunk 2 above)
Comment (samiuelson) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/feedback/ipp/GetIPPFeedbackBannerData.kt]: Good catch, I'll fix that.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListViewModel.kt (hunk 1 above)
Comment (samiuelson) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListViewModel.kt:187]: I've added the unit tests for verifying correct analytics events and props 👍🏼

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/feedback/ipp/GetIPPFeedbackBannerData.kt (hunk 2 above)
Comment (samiuelson) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/feedback/ipp/GetIPPFeedbackBannerData.kt]: This is fixed now. In case of backend errors IPP feedback banner is not shown, and the "Create orders..." banner is shown as usual.

**Comment by samiuelson** [comment]
> Thanks for your work on this @samiuelson. Great work!
> 
> I've smoke tested the app and it seems to be working fine. I couldn't do thorough testing today because of a migraine. I'll be AFK tomorrow and will be back on Friday. But it will be late for your PR merge.
> 
> Since I've already reviewed major part of the PR, I've skimmed through the code and it looks good.
> 
> I'll leave it to @vanessamaartinstt for testing the app.
> 
> I've left a couple of questions to think about. Let me know what you feel about those.

Thank you @AnirudhBhat so much for testing the app and reviewing all the PRs!

I’ve fixed the issues you identified and added unit tests verifying if correct analytics events are tracked.

After manually testing different scenarios myself, I’ve also fixed the behavior in the case no active payments plugins are detected and when the backend (potentially) returns a successful but null response. 

Given this is the last day to merge the PR I’d suggest doing it today (26th of Jan). If there are any disqualifying issues found during the Mob testing we’ll still be able to fix them during the beta testing phase or disable the IPP banner using the feature flag before releasing it.

**List of scenarios fixed directly in this PR:**
1. Merchant has COD enabled but no payments plugin installed. 

In the above ☝🏼 case app was not showing the banner. Now: should show “ipp_not_user” banner.

2. Merchant has COD enabled but the backend returns an error if the payments plugin is not fully configured or (potentially) a null response from `transactions/summary` endpoint. 

In the above ☝🏼 case app was crashing. Now: should show “ipp_not_user” banner.
Basically, it was wrongly assumed that the backend response will always be successful. Now in case of an error response it is logged, the app is not crashing and the IPP banner is shown.

3. Added unit tests to `OrderListViewModelTest` verifying if correct analytics events are tracked when interacting with the banner.

File: WooCommerce/src/test/kotlin/com/woocommerce/android/ui/payments/feedback/ipp/GetIPPFeedbackBannerDataTest.kt [hunk 3]
```diff
@@ -257,20 +257,21 @@ class GetIPPFeedbackBannerDataTest : BaseUnitTest() {
     }
 
     @Test
-    fun `given endpoint returns null response, then should throw exception`() = runBlocking {
+    fun `given endpoint returns null response, then should detect newbie`() = runBlocking {
```
Comment (malinajirka) [inline WooCommerce/src/test/kotlin/com/woocommerce/android/ui/payments/feedback/ipp/GetIPPFeedbackBannerDataTest.kt]: 🔍 Not related to this PR and not important, but just to spread knowledge about this. We have `testBlocking` method which lives in the `BaseUnitTest`. In this case it doesn't change anything, but in general is preferred to ensure consistency in the way we handle coroutines in tests.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivityViewModel.kt [hunk 4]
```diff
@@ -46,6 +46,9 @@ class MainActivityViewModel @Inject constructor(
         launch {
             featureAnnouncementRepository.getFeatureAnnouncements(fromCache = false)
         }
+        prefs.setIPPFeedbackSurveyCompleted(false)
```
Comment (malinajirka) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivityViewModel.kt]: ⚠️ @samiuelson is this intentionally pushed into the repo? Seems like we'd be reseting the banner after each restart of the app, unless I'm missing something.

Comment (samiuelson) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivityViewModel.kt]: My bad 😞 Removed!

Comment (malinajirka) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivityViewModel.kt]: No worries, thanks!

File: WooCommerce/src/test/kotlin/com/woocommerce/android/ui/payments/feedback/ipp/GetIPPFeedbackBannerDataTest.kt (hunk 3 above)
Comment (samiuelson) [reply WooCommerce/src/test/kotlin/com/woocommerce/android/ui/payments/feedback/ipp/GetIPPFeedbackBannerDataTest.kt]: Switched to `testBlocking` in both `GetIPPFeedbackBannerDataTest` and `ShouldShowFeedbackBannerTest`.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivityViewModel.kt (hunk 4 above)
Comment (samiuelson) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivityViewModel.kt]: I've committed the wrong file with [ad72b7b](https://github.com/woocommerce/woocommerce-android/pull/8247/commits/ad72b7bcd79bb49f7f2dadb58d217051af93c7e0). It's reverted with [a534717](https://github.com/woocommerce/woocommerce-android/pull/8247/commits/a5347176603b10d03fe7a913817b26e93f06c952). 

The intended changes are committed with [0046ec7](https://github.com/woocommerce/woocommerce-android/pull/8247/commits/0046ec7c0bc9aee0b88ca395df906c57d8e3e816).

**Review by malinajirka — APPROVED** [review]
Thanks for working on this @samiuelson, great job! 👏 

I've tested several scenarios, including deactivating WCPay and toggling COD and all worked as expected 🏅.
